osu.Game/Screens/SelectV2/BeatmapCarouselFilterGrouping.cs

```diff
@@ -225,18 +233,52 @@ private List<GroupMapping> getGroups(List<CarouselItem> items, FilterCriteria cr
                     return getGroupsBy(b => defineGroupBySource(b.BeatmapSet!.Metadata.Source), items);
 
                 case GroupMode.Collections:
-                    var collections = getCollections?.Invoke() ?? Enumerable.Empty<BeatmapCollection>();
-                    return getGroupsBy(b => defineGroupByCollection(b, collections), items);
+                {
+                    var realm = getRealm?.Invoke();
+
+                    return realm?.Run(r =>
+                    {
+                        var collections = r.All<BeatmapCollection>().AsEnumerable();
+                        return getGroupsBy(b => defineGroupByCollection(b, collections), items);
+                    }) ?? new List<GroupMapping>();
+                }
 
                 case GroupMode.MyMaps:
                     return getGroupsBy(b => defineGroupByOwnMaps(b, criteria.LocalUserId, criteria.LocalUserUsername), items);
 
+                case GroupMode.RankAchieved:
+                {
+                    var realm = getRealm?.Invoke();
+
+                    var topRankMapping = new Dictionary<BeatmapInfo, ScoreRank>(items.Count);
+
+                    return realm?.Run(r =>
+                    {
+                        var allLocalScores = r.All<ScoreInfo>()
+                                              .Filter($"{nameof(ScoreInfo.User)}.{nameof(RealmUser.OnlineID)} == $0"
+                                                      + $" && {nameof(ScoreInfo.BeatmapInfo)}.{nameof(BeatmapInfo.Hash)} == {nameof(ScoreInfo.BeatmapHash)}"
+                                                      + $" && {nameof(ScoreInfo.Ruleset)}.{nameof(RulesetInfo.ShortName)} == $1"
+                                                      + $" && {nameof(ScoreInfo.DeletePending)} == false", criteria.LocalUserId, criteria.Ruleset?.ShortName)
+                                              .OrderByDescending(s => s.TotalScore)
+                                              .ThenBy(s => s.Date);
```

Comment:
Date has limited precision. Online score id should be preferred whenever available.

Reply:
Also carried from https://github.com/ppy/osu/blob/cb354685ca323f4e429e60617077ed5a5175ee3d/osu.Game/Screens/Select/Carousel/TopLocalRank.cs#L78-L77, I can switch it to fall back to ID but we should probably update that code as well? If so, should I bundle it in this PR or in a separate one.